pkg/scheduler/dependency_graph.go

```diff
@@ -502,26 +588,109 @@ func (sched *Scheduler) BuildDependencyGraph(options interfaces.DependencyGraphO
 		for _, block := range blocks {
 			for _, entry := range block {
 				key := entry.scheduledResource.Key()
-				existingSr := depGraph.graph[key]
+				existingSr := rootContext.graph.graph[key]
 				if existingSr == nil {
-					log.Printf("Adding resource %s to the dependency graph flow %s", key, options.FlowName)
-					depGraph.graph[key] = entry.scheduledResource
+					if !silent {
+						log.Printf("Adding resource %s to the dependency graph flow %s", key, flow.Name)
+					}
+					rootContext.graph.graph[key] = entry.scheduledResource
 				} else {
 					sched.updateContext(existingSr.Context, entry.parentContext, entry.dependency)
 					existingSr.Requires = append(existingSr.Requires, entry.scheduledResource.Requires...)
 					existingSr.RequiredBy = append(existingSr.RequiredBy, entry.scheduledResource.RequiredBy...)
+					existingSr.usedInReplicas = append(existingSr.usedInReplicas, entry.scheduledResource.usedInReplicas...)
 					for metaKey, metaValue := range entry.scheduledResource.Meta {
 						existingSr.Meta[metaKey] = metaValue
 					}
 				}
 			}
 		}
 	}
+	return nil
+}
 
-	for _, value := range depGraph.graph {
-		value.RequiredBy = unique(value.RequiredBy)
-		value.Requires = unique(value.Requires)
+func (sched *Scheduler) composeFinalizer(construction, destruction *DependencyGraph, replicas []client.Replica) func() {
+	replicaMap := map[string]client.Replica{}
+	for _, replica := range replicas {
+		replicaMap[replica.ReplicaName()] = replica
 	}
 
-	return depGraph, nil
+	var destructors []func() bool
+	var failed chan *ScheduledResource
+
+	var resourceCount int
+	for _, depGraph := range [2]*DependencyGraph{construction, destruction} {
+		for _, resource := range depGraph.graph {
+			release := true
+			for _, replicaName := range resource.usedInReplicas {
+				if _, found := replicaMap[replicaName]; !found {
+					release = false
+					break
+				}
+			}
+			if release {
+				resourceCopy := resource
+				f := func() bool {
```

Comment:
This function can be returned from a named function `getDestructor` which has parameters `resourceCopy`, failed`.